### PR TITLE
CPU_PROFILE Configuration In Script, Better Support For VBox Installs On A Different Drive Than C:, And CPU Recognition For The Beginners

### PR DESCRIPTION
**Pull request by TheBotlyNoob**
Added the automatic CPU_PROFILE config variable so it is a lot simpler to change the cpu profile. And its easier to make more VMs with the same configuration.

### CODE REVIEW

File: macos-guest-virtualbox.sh [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #               tesseract-ocr >= 4
 
 function set_variables() {
+cpu_profile_on=false
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:17]: Not required, please remove this line

File: macos-guest-virtualbox.sh [hunk 2]
```diff
@@ -22,7 +23,8 @@ storage_format="vdi"             # VM disk image file format, "vdi" or "vmdk"
 cpu_count=2                      # VM CPU cores, minimum 2
 memory_size=4096                 # VM RAM in MB, minimum 2048
 gpu_vram=128                     # VM video RAM in MB, minimum 34, maximum 128
-resolution="1280x800"            # VM display resolution
+resolution="1280x800"            # VM display resolution 
+cpu_profile="Intel Core i7-3960X" && cpu_profile_on=true #If having issues on the boot screen try uncommenting this
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:27]: Default CPU profile should be `host`

File: macos-guest-virtualbox.sh [hunk 3]
```diff
@@ -93,6 +95,9 @@ pad_to_33_chars "cpu_count=${cpu_count}"                       "# VM CPU cores,
 pad_to_33_chars "memory_size=${memory_size}"                   "# VM RAM in MB, minimum 2048"
 pad_to_33_chars "gpu_vram=${gpu_vram}"                         "# VM video RAM in MB, minimum 34, maximum 128"
 pad_to_33_chars "resolution=\"${resolution}\""                 "# VM display resolution"
+if [ ${cpu_profile_on} = true ] ; then
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:98]: if statement not required

File: macos-guest-virtualbox.sh [hunk 4]
```diff
@@ -673,6 +678,10 @@ VBoxManage setextradata "${vm_name}" \
   "ourhardworkbythesewordsguardedpleasedontsteal(c)AppleComputerInc"
 VBoxManage setextradata "${vm_name}" \
  "VBoxInternal/Devices/smc/0/Config/GetKeyFromRealSMC" 0
+if [ ${cpu_profile_on} = true ] ; then
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:681]: The `--cpu-profile "${cpu_profile}` option should be added to the existing `modifyvm` command when configuring the virtual machine.

**Comment by TheBotlyNoob** [comment]
I have done all of your edits.

**Comment by myspaghetti** [comment]
Thanks.

Sorry, but it just occurred to me: if no CPU profile is required then this change does no harm, but if it *is* required then this change does no good since the user will have to manually choose the CPU profile anyway. This just adds an extra step. Instead of running `VBoxManage modifyvm ...` and then `./macos-guest-virualbox.sh populate...`, the user has to edit the script then run `./macos-guest-virualbox.sh configure_vm populate...`

On the other hand, if the user keeps the script for future virtual machines, then the `configure_vm` step will correctly configure their CPU profile.

Personally, I think running a terminal command six times is easier than editing a shell script six times, one less state change each time.

I'm not closing the pull request yet, but I have to think about it.

**Comment by TheBotlyNoob** [comment]
Okay, all I know is that most of my CPU's are incompatible so I just leave it on Intel Core i7-3960X for convenience.

**Comment by TheBotlyNoob** [comment]
I mean maybe you could just link my respiratory in the CPUID section. This is just for convenience I really don't care if it get merged or not.

**Comment by Yethal** [comment]
@myspaghetti If the user knows for a fact they're going to change the cpu profile (so far all Ryzen systems required it for example) then they can edit the profile before the first run and have a working VM on the first try

**Comment by myspaghetti** [comment]
That's true. I believe only advanced users know beforehand which CPU profile or CPUID they require, so I'm not sure how much of a convenience it is to add a `cpu_profile` variable when all it does is add `--cpu-profile "${cpu_profile} ` to line 645. It's definitely *slightly more* convenient to edit the variable at the top of the script than hunt around for line 645.

I'm considering it.

**Comment by myspaghetti** [comment]
This is good. I'll review/merge it soon.

**Comment by TheBotlyNoob** [comment]
@myspaghetti Just a quick question, why is the max VRAM 128? Because the command VBoxManage.exe" modifyvm "macOS" --vram 256 works like a charm?

**Comment by TheBotlyNoob** [comment]
Ok, after a short decision I have decided that macOS really isn't my cup of tea. So I will keep my fork maintained and maybe even make more features, but im never going to use it (except for bug fixes).

File: macos-guest-virtualbox.sh [hunk 5]
```diff
@@ -217,45 +232,65 @@ if [[ -n "$(cygcheck -V 2>/dev/null)" ]]; then
...
-            }
-            echo "Found VBoxManage"
-        else
-            echo "Please make sure VirtualBox version 5.2 or higher is installed, and that"
-            echo "the path to the VBoxManage.exe executable is in the PATH variable, or assign"
-            echo "in the script the full path including the name of the executable to"
-            echo -e "the variable ${highlight_color}cmd_path_VBoxManage${default_color}"
-            exit
-        fi
+    	# drives A and B are last because they are reserved for floppy drives
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:235]: Please do this more cleanly

**Comment by myspaghetti** [comment]
I see there have been further changes; they're all pretty positive but some cleanup is required, for example instead of specifying `"C" "D" "E" ..."` there should be a more expedient way to cycle through the alphabet.

**Comment by myspaghetti** [comment]
>Just a quick question, why is the max VRAM 128? Because the command VBoxManage.exe modifyvm "macOS" --vram 256 works like a charm?

There might have been some old versions of VirtualBox that didn't support this. I'll have to check.

**Comment by TheBotlyNoob** [comment]
It seems to have been supported since 2015 [Here is where I founs the command](https://askubuntu.com/questions/587083/virtualbox-how-to-increase-video-memory)

**Comment by myspaghetti** [comment]
It's true that VirtualBox has supported 256MB of VRAM since about 2010, but not across 5.2 and 6.0 when I first wrote the script. I guess I'll test it again now across the latest versions of 5.2, 6.0, and 6.1.

File: macos-guest-virtualbox.sh [hunk 6]
```diff
@@ -217,45 +232,66 @@ if [[ -n "$(cygcheck -V 2>/dev/null)" ]]; then
...
+	declare -a driveletters={C..Z} && driveletters+=("A" "B")
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:236]: `declare -a driveletters={C..Z} ={C..Z}` seems to expand `driveletters={C..Z}` to `Z` only on bash version `4.4.12(3)-release`

File: macos-guest-virtualbox.sh [hunk 7]
```diff
@@ -217,45 +232,66 @@ if [[ -n "$(cygcheck -V 2>/dev/null)" ]]; then
...
+	# f is a counter
+	declare -i f=0
+	for i in ${driveletters[@]}; do
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:239]: Why not skip the counter and declaration and just `for i in {C..Z} A B`

File: macos-guest-virtualbox.sh [hunk 8]
```diff
@@ -320,6 +356,18 @@ if [[ -z "$(dmg2img -d 2>/dev/null)" ]]; then
...
 
+# get CPU Information from /proc/cpuinfo
+modelname=$(cat /proc/cpuinfo | grep 'model name' | head -1)
+# check if cpu_profile contains something other than host, as to not disturb testers trying a different cpu_profile.
+if [[ ${cpu_profile} != "host" ]]; then
+	echo "cpu_profile Is $cpu_profile, Other Than host, So Not Changing Anything"
+elif grep Intel <<<"$modelname" > /dev/null; then
+	echo "Intel CPU Detected, So Not Changing Anything"
+elif grep "AMD" <<<"$modelname" > /dev/null; then
+	echo "AMD CPU Detected, So Changing cpu_profile To: Intel Core i7-3960X"
```
Comment (myspaghetti) [inline macos-guest-virtualbox.sh:367]: As stated elsewhere, I will not maintain a database of CPUs and their matching profiles.

**Comment by myspaghetti** [comment]
I'm closing this pull request. Some of the contributions are good, others are excluded by design.

* 256MB VRAM: I (or someone else) will have to test this across VirtualBox 5.2, 6.0, and 6.1 on Windows, Linux, and macOS.
* Checking drive letters other than `C`: beneficial, but not implemented cleanly, and if you have VirtualBox installed on a non-default location there's already an error message asking you to specify where it is, assuming that if the user is advanced enough for non-default installation paths they are advanced enough to specify where they installed the software
* CPU profile database: this is excluded by design.

The VRAM will most likely be implemented (when I or someone else test it) and can be submitted in a separate pull request. The drive letters are fine and can be submitted in a separate pull request but the issue affects very few people, and the people it does affect are likely to have installed VirtualBox outside of the default "Program Files" path anyway. The CPU database (even with a data set of 1 entry) will not be implemented.

**Pull request closed without merging**